### ISSUE TITLE
Proposal block can never be committed

### ISSUE THREAD

**Issue by kwunyeung** [strangelove-ventures/horcrux#30]
With `horcrux` setup correctly, the validator's proposal block will always get the following info and the block will never be committed, resulting selecting a new proposer and slow down the speed of chain.

```
INF commit is for a block we do not know about; set ProposalBlock=nil commit=C52CB815E75718494DA41B0D49006B7E90597BE39B8A3FF17215596E148F0FC2 commit_round=1 height=2900785 module=consensus proposal=
```
The following is the log from the signer. We can see that another precommit is voted at `round=1`.

```
Nov 15 05:06:16 desmos-m-signer-3-ln-sg horcrux[153734]: I[2021-11-15|05:06:16.569] Signed proposal                              module=validator node={validator_address} height=2900785 round=0 type=SIGNED_MSG_TYPE_PROPOSAL
Nov 15 05:06:17 desmos-m-signer-3-ln-sg horcrux[153734]: I[2021-11-15|05:06:17.850] Signed vote                                  module=validator node={validator_address} height=2900785 round=0 type=SIGNED_MSG_TYPE_PREVOTE
Nov 15 05:06:19 desmos-m-signer-3-ln-sg horcrux[153734]: I[2021-11-15|05:06:19.131] Signed vote                                  module=validator node={validator_address} height=2900785 round=0 type=SIGNED_MSG_TYPE_PRECOMMIT
Nov 15 05:06:20 desmos-m-signer-3-ln-sg horcrux[153734]: I[2021-11-15|05:06:20.433] Signed vote                                  module=validator node={validator_address} height=2900785 round=1 type=SIGNED_MSG_TYPE_PRECOMMIT
```

**Comment by jackzampolin** [comment]
Try increasing the timeout between the signers in the cluster from 1000ms to 1500ms

**Comment by kwunyeung** [comment]
@jackzampolin we have tried that even increased the timeout to `2000ms` but the validator still can't propose a valid block. If the signer's timeout is `1500ms`, should the proposal timeout change from `3s` to `4.5s` for a 3 signers setup? I see that the proposal is signed but when it cannot add votes in round 0.

**Comment by jackzampolin** [comment]
Oh so I've figured out whats going on. Standard architecture for `horcrux` consumers is 3 sentries connected to 3 signers in a 2/3 configuration. The following are logs from the 3 signers during a sample proposal:

### Node 3:
```
Dec 14 22:21:48 hub-mainnet-shard-3 signer[3526778]: E[2021-12-14|22:21:48.682] Failed to sign proposal                      module=validator address=tcp://10.138.0.14:1234 error="Mismatched data" proposal="type:SIGNED_MSG_TYPE_PROPOSAL height:8696433 pol_round:-1 block_id:<hash:\"H\\315e8P>H\\217qi\\231\\221\\2420\\350\\237\\017L&\\373\\354\\370\\364\\346\\203\\246\\252\\347\\225\\322\\001u\" part_set_header:<total:1 hash:\"\\213W.\\352\\032\\035\\361\\215\\371\\247\\374\\205\\037\\224*\\350\\010\\202\\026\\256\\2617\\034\\356\\322+\\304y@m?^\" > > timestamp:<seconds:1639520508 nanos:348674158 >
```
### Node 2: 
```
Dec 14 22:21:48 hub-mainnet-shard-2 signer[1051]: ERROR Sign RPC error -32603 - Internal error: Mismatched data
Dec 14 22:21:48 hub-mainnet-shard-2 signer[1051]: E[2021-12-14|22:21:48.688] Failed to sign proposal                      module=validator address=tcp://10.182.0.2:1234 error="Mismatched data" proposal="type:SIGNED_MSG_TYPE_PROPOSAL height:8696433 pol_round:-1 block_id:<hash:\"M$y\\023}u\\206z<=\\276\\371\\372\\364R\\021\\324Y\\202\\347m\\257\\010\\300Ri\\317N\\276\\226\\234f\" part_set_header:<total:1 hash:\"\\202\\270,{)\\200P#\\214l\\330\\214j\\376\\376\\365\\263\\235R5\\177\\376\\340B=\\272\\336\\004\\203\\200\\301\\350\" > > timestamp:<seconds:1639520508 nanos:437183246 > "
```

### Node 1: 
```Dec 14 22:21:48 hub-mainnet-shard-1 signer[2828304]: ERROR Sign RPC error -32603 - Internal error: Mismatched data
Dec 14 22:21:48 hub-mainnet-shard-1 signer[2828304]: ERROR Sign RPC error -32603 - Internal error: Mismatched data
Dec 14 22:21:48 hub-mainnet-shard-1 signer[2828304]: E[2021-12-14|22:21:48.672] Failed to sign proposal                      module=validator address=tcp://10.128.0.20:1234 error="Mismatched data" proposal="type:SIGNED_MSG_TYPE_PROPOSAL height:8696433 pol_round:-1 block_id:<hash:\"\\022L\\022\\307\\321}\\227\\346\\361]#\\247\\377\\231.\\207\\312\\274\\220\\270i\\210NU1\\204P\\\"]&U\\177\" part_set_header:<total:1 hash:\"8\\232\\311\\271\\340\\207m\\372c>.\\024\\370\\252\\266\\003\\263\\237)\\362!\\323v\\251\\306xN\\022\\217\\314\\354\\251\" > > timestamp:<seconds:1639520508 nanos:424605029 > "
```

What happens is that each signer, connected to a different node, gets a different block. This sets the `HRS` (height, round, step) for the signer and effectively poisons that signer to signing another proposal. Its unclear exactly how to resolve this

**Comment by kwunyeung** [comment]
@jackzampolin have you ever been proposing blocks successfully on any Stargate chains?

**Comment by cmwaters** [comment]
It's still not clear to me what's happening. Is this only in the case when the validator is proposing a block right? So all three nodes try to propose different blocks and the signer doesn't end up signing any of them so the validator ends up never proposing a block

**Comment by jackzampolin** [comment]
@cmwaters yup. As the proposals come into the signer nodes they set their high watermark files. Because each proposal contains different transactions, when they gossip their proposal to the adjacent nodes they are rejected.

**Comment by cmwaters** [comment]
But wouldn't the other nodes still accept the first proposal they see (and only reject the latter ones)?

**Comment by jackzampolin** [comment]
@cmwaters No, because they see a proposal, they accept it uncritically and think it is _the_ proposal. This happens on each node and they reject subsequent requests to sign proposals from other nodes because they already have their Height Round Step high watermark set.

**Comment by jackzampolin** [comment]
So I think the fix here is to elect a leader in the signer cluster. Simple way to do this would be to say that for each proposal `leader := (blockHeight%numSigners)+1`. The other two nodes would, upon seeing a proposal wait for a configurable amount of time before signing a proposal they had. This would fix all cases except downed leader which is a huge improvement.

**Comment by jackzampolin** [comment]
Actual fix here is #36 (i.e. raft)
